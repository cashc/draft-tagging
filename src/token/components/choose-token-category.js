import { h } from 'react-hyperscript-helpers';
import { connect } from 'react-redux';

import { setSelectedTokenCategory } from '../action-creators';
import { DELETE_TOKEN_CATEGORY } from '../constants';
import { getCategoryItemsSorted } from '../util';

function buildCategoryActions(onClick) {
  const categoryItems = getCategoryItemsSorted();
  const categoryActions = categoryItems.map((categoryItem) =>
    createCategoryItemAction(categoryItem, onClick),
  );

  const remover = {
    type: 'CONTEXT_MENU_ITEM_TYPE',
    label: 'Remove',
    name: DELETE_TOKEN_CATEGORY,
    onClick: () => onClick(DELETE_TOKEN_CATEGORY),
  };
  return [...categoryActions, remover];
}

function createCategoryItemAction(categoryItem, onClick) {
  const itemType = 'CONTEXT_MENU_ITEM_TYPE';
  const { categoryType, categoryTitle } = categoryItem;
  return {
    type: itemType,
    label: categoryTitle,
    name: categoryType,
    onClick: () => onClick(categoryType),
  };
}

export function ChooseTokenCategory({
  // clickPosition,
  // close = () => {},
  handleChooseTokenCategory,
}) {
  const actions = buildCategoryActions(handleChooseTokenCategory);

  return h('div', actions.map((a) => h('div', a.categoryTitle)));
}

function mapDispatchToProps(dispatch, { selectedToken, close }) {
  return {
    handleChooseTokenCategory: (category) => {
      dispatch(setSelectedTokenCategory({ selectedToken, category }));
      close();
    },
  };
}

export const ChooseTokenCategoryConn = connect(null, mapDispatchToProps)(
  ChooseTokenCategory,
);
