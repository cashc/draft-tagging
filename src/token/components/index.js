export { SearchBarConn } from './search-bar';
export { SearchTokenChipConn } from './search-token-chip';
