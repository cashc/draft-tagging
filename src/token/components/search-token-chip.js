import { h } from 'react-hyperscript-helpers';
import styled from 'styled-components';
import { connect } from 'react-redux';

import { getTokensForWordsSorted } from '../selectors';

const SearchTokenChipContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  flex-direction: column;
  padding: 5px 0 18px 31px;
  background-color: ${(props) => props.theme.backgroundPrimary};
`;

const HelpTextDiv = styled.div`
  margin-bottom: 4px;
  font-size: 13px;
  letter-spacing: 0.1px;
  text-align: left;
  color: ${(props) => props.theme.foregroundPrimary};
`;

const WordRowDiv = styled.div`
  display: flex;
  align-items: center;
  margin: 6px 0;
`;

const WordDiv = styled.div`
  margin-right: 4px;
  font-size: 13px;
  font-weight: 600;
  line-height: 1.54;
  letter-spacing: 0.1px;
  text-align: left;
  color: ${(props) => props.theme.foregroundPrimary};
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
`;

const Chip = styled.div`
  display: flex;
  align-items: center;
  min-height: 22px;
  margin: 0 3px;
  padding-right: 2px;
  color: ${(props) => props.theme.foregroundLink};
  border-radius: 12px;
  box-shadow: 1px 1px 10px ${({ theme }) => theme.boxShadow};
  background-color: ${(props) => props.theme.backgroundDisabledInput};
  cursor: pointer;

  &:hover {
    text-shadow: 0 0 1px ${(props) => props.theme.infoBackground};
  }
`;

const Category = styled.span`
  padding: 0 1px 0 6px;
`;

const DropdownTriangle = styled.div`
  margin: 0px 3px -6px 2px;
  width: 0;
  height: 0;
  border: 4px solid transparent;
  border-top: 4px solid ${(props) => props.theme.foregroundLink};
`;

export const QueryHitsDiv = styled.div`
  font-size: 13px;
  letter-spacing: 0.1px;
  text-align: left;
  color: ${(props) => props.theme.foregroundSecondary};
  margin-top: 5px;
`;

export const SearchTokenChip = ({
  tokensForWords = [],
  chooseTokenCategory = () => {},
}) => {
  const tokensForWordsRows = tokensForWords.map((tokensForWord) => {
    const word = tokensForWord[0].token;
    const tokens = tokensForWord.map((token) =>
      TokenChip(token, { chooseTokenCategory }),
    );
    return h(WordRowDiv, [h(WordDiv, `${word}:`), ...tokens]);
  });

  return h(SearchTokenChipContainer, [
    h(HelpTextDiv, 'Highlight a selection you want to tag and then click on it to choose a tag.'),
    ...tokensForWordsRows,
  ]);
};

const TokenChip = (token, { chooseTokenCategory }) => {
  const { categoryTitle } = token;
  return h(
    Chip,
    {
      onClick: (event) => chooseTokenCategory(event, token),
    },
    [h(Category, categoryTitle), h(DropdownTriangle)],
  );
};

const mapStateToProps = (state) => ({
  tokensForWords: getTokensForWordsSorted(state),
});

export const mapDispatchToProps = () => ({
  chooseTokenCategory: (event, selectedToken) => {
    console.log('HIT', selectedToken);
  },
});

export const SearchTokenChipConn = connect(mapStateToProps, mapDispatchToProps)(
  SearchTokenChip,
);
