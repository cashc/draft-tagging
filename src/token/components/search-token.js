import { Component } from 'react';
import { connect } from 'react-redux';
import { h } from 'react-hyperscript-helpers';
import styled from 'styled-components';

import { InjectSVG } from '../../ui';
import { NEW_TOKEN_CATEGORY } from '../constants';
import { openModal, setSelectedToken } from '../action-creators';

const SearchBarToken = styled.span`
  border-radius: 2px;
  background-color: ${(props) => props.theme.infoBackground};

  > .handlebar {
    display: none;
    position: absolute;

    > svg {
      width: 6px;
      height: 21px;
    }

    > svg * {
      fill: ${(props) => props.theme.brandPrimary};
    }
  }

  &:hover > div {
    display: inline-block;
  }

  > .left {
    margin-left: -2px;
  }

  > .right {
    margin-top: 2px;
    margin-left: -4px;
  }
`;

const NewSearchBarToken = SearchBarToken.extend`
  cursor: pointer;
  border: 1px solid;
  background-color: none;
`;

const DraggableHandleBase = styled.div`
  position: absolute;
  height: 24px;
  width: 9px;
  cursor: pointer;
  user-select: none;
  display: none;
`;

const DraggableHandleLeft = DraggableHandleBase.extend`
  margin-left: -2px;
`;

const DraggableHandleRight = DraggableHandleBase.extend`
  margin-top: 2px;
  margin-left: -4px;
`;

type Props = {
  token: Object,
  children: Array<Object>,
  chooseTokenCategory: Function,
  onHandleMouseDown: Function,
};

export class SearchToken extends Component {
  props: Props;

  static defaultProps = {
    token: {},
    children: [],
    chooseTokenCategory: () => {},
    onHandleMouseDown: () => {},
  };

  onStart = (ev, isLeft) => {
    ev.stopPropagation();
    const { token, onHandleMouseDown } = this.props;
    onHandleMouseDown(isLeft, token);
  };

  render() {
    const { token, children, chooseTokenCategory } = this.props;

    if (token.category === NEW_TOKEN_CATEGORY) {
      return h(
        NewSearchBarToken,
        { onClick: (ev) => chooseTokenCategory(ev, token) },
        children,
      );
    }

    return h(SearchBarToken, { onClick: (ev) => stopEvent(ev) }, [
      h(InjectSVG, {
        path: 'text-handle-start.svg',
        containerClassName: 'handlebar left',
      }),
      h(DraggableHandleLeft, { onMouseDown: (ev) => this.onStart(ev, true) }),
      ...children,
      h(InjectSVG, {
        path: 'text-handle-end.svg',
        containerClassName: 'handlebar right',
      }),
      h(DraggableHandleRight, { onMouseDown: (ev) => this.onStart(ev, false) }),
    ]);
  }
}

const stopEvent = (ev) => {
  ev.preventDefault();
  ev.stopPropagation();
};

const mapDispatchToProps = (dispatch) => ({
  chooseTokenCategory: (ev, selectedToken) => {
    stopEvent(ev);
    dispatch(setSelectedToken({ selectedToken }));
    dispatch(openModal({ selectedToken }));
  },
});

export const SearchTokenConn = connect(null, mapDispatchToProps)(SearchToken);
