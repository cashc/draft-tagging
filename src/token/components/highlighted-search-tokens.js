import { Component } from 'react';
import {
  Editor,
  EditorState,
  CompositeDecorator,
  SelectionState,
  Modifier,
  convertToRaw,
  convertFromRaw,
} from 'draft-js';
import lodash from 'lodash';
const { debounce } = lodash;
import { h } from 'react-hyperscript-helpers';
import styled from 'styled-components';

import { queryHasToken } from '../util';
import { NEW_TOKEN_CATEGORY } from '../constants';
import { SearchTokenConn } from './search-token';

const TOKEN_ENTITY_TYPE = 'TOKEN';
const QUERY_TOKEN_DEBOUNCE = 1000;
const HIGHLIGHT_SELECTION_DEBOUNCE = 300;
const HANDLED = 'handled';
const DRAFT_JS_DIV = 'public-DraftStyleDefault-block';
const RIGHT_HANDLE = 'draggable-handle right-handle';
const LEFT_HANDLE = 'draggable-handle left-handle';
const PLACEHOLDER_TEXT = 'Search Connections';

const SearchContainer = styled.div`
  display: flex;
  min-height: 48px;
  border: 2px solid ${({ theme }) => theme.borderPrimary};
  border-radius: 7px;
`;

const SearchInput = styled.div`
  width: calc(100% - 40px);
  overflow: hidden;
  padding: 12px 32px 13px 32px;
  font-size: 18px;
  letter-spacing: -0.5px;
  text-align: left;
  color: ${(props) => {
    if (props.hasText) return props.theme.foregroundPrimary;
    return props.theme.foregroundPlaceholder;
  }};
`;

const EditorParent = styled.div``;

export class HighlightedSearchTokens extends Component {
  static defaultProps = {
    query: '',
    tokens: [],
    handleSearch: () => {},
    updateToken: () => {},
    disableTyping: false,
  };

  state = {
    editorState: EditorState.createEmpty(editorDecorators),
    displayTokens: false,
    tokenBeingResized: null,
    resizedToken: null,
  };

  componentWillMount() {
    const { editorState } = this.state;
    const { query } = this.props;

    const block = { text: query };
    const newContentState = convertFromRaw({ blocks: [block], entityMap: {} });
    const newEditorState = EditorState.push(editorState, newContentState);

    this.onChange(newEditorState);
  }

  componentWillReceiveProps() {
    this.highlightTokensTrailing();
  }

  componentDidMount() {
    document.addEventListener('selectionchange', this.onSelectionChangeHandle);
    if (this.element)
      this.element.addEventListener('mouseup', this.onMouseUpHandle);
  }

  componentWillUnmount() {
    document.removeEventListener(
      'selectionchange',
      this.onSelectionChangeHandle,
    );

    this.highlightTokensTrailing.cancel();

    if (this.element)
      this.element.removeEventListener('mouseup', this.onMouseUpHandle);
  }

  highlightTokens = (displayTokens, newEditorState = null) => {
    const { editorState } =
      newEditorState === null ? this.state : newEditorState;
    const { tokens } = this.props;
    const text = getTextFromEditorState(editorState);

    const clearedEditorState = clearEntities(editorState);
    if (!shouldRenderEntities(text, tokens, displayTokens)) {
      const clearedFocusedEditorState = EditorState.acceptSelection(
        clearedEditorState,
        editorState.getSelection(),
      );
      this.setState({ editorState: clearedFocusedEditorState });
      return clearedFocusedEditorState;
    }

    const highlightedEditorState = applyEntitiesToTokens(
      clearedEditorState,
      tokens,
      { onHandleMouseDown: this.onHandleMouseDown },
    );

    const highlightedFocusedEditorState = EditorState.acceptSelection(
      highlightedEditorState,
      editorState.getSelection(),
    );

    this.setState({ editorState: highlightedFocusedEditorState });

    return highlightedFocusedEditorState;
  };

  // triggered on the trailing edge to allow new props to be set before calling
  highlightTokensTrailing = debounce(() => {
    const { displayTokens } = this.state;
    this.highlightTokens(displayTokens);
  }, 0);

  onChange = (newEditorState) => {
    const { handleSearch, query, disableTyping } = this.props;
    const { editorState } = this.state;

    const text = getTextFromEditorState(editorState);
    const newText = getTextFromEditorState(newEditorState);

    const textIsIdentical = newText === text && newText === query;

    if (disableTyping) {
      if (text !== '' && newText !== text) {
        return;
      }
    }

    if (textIsIdentical) {
      const { start, end } = getSelectionStartAndEnd(newEditorState);
      // check to see if anchor and focus are in the same position
      if (start !== end) {
        this.highlightSelectionOnChange(newEditorState);
      } else {
        this.highlightTokensTrailing();
      }

      // need to set state immediately to see selection range highlighting
      this.setState({ editorState: newEditorState });
      return;
    }

    // if you select something, press delete immediately, cancel before painting new token
    this.highlightSelectionOnChange.cancel();

    const clearedEditorState = clearEntities(newEditorState);
    this.setState({ editorState: clearedEditorState, displayTokens: false });

    if (newText !== query) handleSearch(newText);
    this.highlightTokensOnChange();
  };

  highlightTokensOnChange = debounce(() => {
    this.setState({ displayTokens: true });
    this.highlightTokens(true);
  }, QUERY_TOKEN_DEBOUNCE);

  highlightSelection = (newEditorState) => {
    const { displayTokens } = this.state;
    const editorState = this.highlightTokens(displayTokens, {
      editorState: newEditorState,
    });

    const { start, end } = getSelectionStartAndEnd(editorState);
    const text = getTextFromEditorState(editorState);
    const token = text.substr(start, end - start);
    const newTokenSelection = {
      token,
      start,
      end,
      category: NEW_TOKEN_CATEGORY,
    };

    const selectedEditorState = applyEntityToToken(
      editorState,
      newTokenSelection,
    );
    const focusedEditorState = EditorState.acceptSelection(
      selectedEditorState,
      newEditorState.getSelection(),
    );

    return focusedEditorState;
  };

  highlightSelectionOnChange = debounce((editorState) => {
    this.setState({ editorState: this.highlightSelection(editorState) });
  }, HIGHLIGHT_SELECTION_DEBOUNCE);

  onHandleMouseDown = (isLeft, token) => {
    this.setState({
      tokenBeingResized: token,
      resizedToken: { ...token, isLeft },
    });
  };

  onMouseUpHandle = () => {
    const { tokenBeingResized, resizedToken } = this.state;
    const { updateToken } = this.props;

    if (tokenBeingResized === null) return;

    // dispatch action setting resizedToken
    updateToken(tokenBeingResized, resizedToken);
    this.setState({ tokenBeingResized: null });
  };

  onSelectionChangeHandle = () => {
    const { tokenBeingResized, resizedToken } = this.state;
    if (tokenBeingResized === null) return;

    const selection = window.getSelection();
    const focus = getSelectionFocus(selection, this.element);
    if (isNaN(focus)) return;

    const { start, end } = resizedToken;
    // only update token if focus has moved from token start/end
    if (start === focus || end === focus) return;
    // make selection at single point, instead of a range
    selection.collapse(selection.focusNode, selection.focusOffset);

    const { query } = this.props;
    const { editorState } = this.state;

    const newResizedToken = getResizedToken(query, resizedToken, focus);

    const newEditorState = updateTokenEntity(
      editorState,
      resizedToken,
      newResizedToken,
      focus,
    );

    this.setState({
      resizedToken: newResizedToken,
      editorState: newEditorState,
    });
  };

  reset = () => {
    const { handleSearch } = this.props;
    handleSearch('');

    this.highlightTokensOnChange.cancel();
    const editorState = EditorState.createEmpty(editorDecorators);
    const selection = editorState.getSelection();
    const focusedEditorState = EditorState.forceSelection(
      editorState,
      selection,
    );
    this.setState({ editorState: focusedEditorState });
  };

  searchOnClick = () => {
    const { editorState } = this.state;
    const { handleSearch } = this.props;
    const text = getTextFromEditorState(editorState);
    handleSearch(text, false);
  };

  editorHandleReturn = () => {
    this.searchOnClick();
    return HANDLED;
  };

  setElementRef = (c) => {
    if (!c) return;
    const children = c.getElementsByClassName(DRAFT_JS_DIV);
    if (children.length > 0) this.element = children[0];
  };

  render() {
    const { editorState } = this.state;
    const hasText = editorState.getCurrentContent().hasText();
    return h(SearchContainer, [
      h(SearchInput, { hasText }, [
        h(
          EditorParent,
          {
            innerRef: (c) => this.setElementRef(c),
          },
          [
            h(Editor, {
              editorState,
              onChange: this.onChange,
              handleReturn: this.editorHandleReturn,
              placeholder: PLACEHOLDER_TEXT,
            }),
          ],
        ),
      ]),
    ]);
  }
}

export const editorDecorators = new CompositeDecorator([
  { strategy: tokenStrategy, component: Token },
]);

function Token(props) {
  const { entityKey, contentState, children } = props;
  const { data } = contentState.getEntity(entityKey);

  return h(SearchTokenConn, { children, ...data });
}

// create a token from the query with the provided start and end indicies
export const createToken = (startIndex, endIndex, query, category) => {
  const start = startIndex >= 0 ? startIndex : 0;
  const end = endIndex < query.length ? endIndex : query.length;
  const length = end - start;
  const token = query.substr(start, length);
  return { start, end, token: token, category };
};

export const getResizedToken = (query, token, focus) => {
  const { isLeft } = token;
  let focusWithinEntityRange;
  let start;
  let end;

  if (isLeft) {
    focusWithinEntityRange = focus < token.end;
    start = focusWithinEntityRange ? focus : token.end;
    end = focusWithinEntityRange ? token.end : focus;
  } else {
    focusWithinEntityRange = focus > token.start;
    start = focusWithinEntityRange ? token.start : focus;
    end = focusWithinEntityRange ? focus : token.start;
  }

  const resizedToken = createToken(start, end, query, token.category);
  return { ...resizedToken, isLeft: focusWithinEntityRange ? isLeft : !isLeft };
};

// Returns editor state with resized token at new start and end indicies.
export const updateTokenEntity = (
  editorState,
  currentToken,
  resizedToken,
  focus,
) => {
  let contentState = editorState.getCurrentContent();
  const block = contentState.getFirstBlock();
  const blockKey = block.getKey();
  const selectionState = SelectionState.createEmpty(blockKey);
  const entityKey = block.getEntityAt(currentToken.start);

  if (
    currentToken.start < resizedToken.start ||
    currentToken.end > resizedToken.end
  ) {
    const tokenSelection = selectionState.merge({
      anchorOffset: currentToken.start,
      focusOffset: currentToken.end,
    });
    const clearedContentState = Modifier.applyEntity(
      contentState,
      tokenSelection,
      null,
    );
    contentState = clearedContentState;
  }

  const resizedTokenSelection = selectionState.merge({
    anchorOffset: resizedToken.start,
    focusOffset: resizedToken.end,
  });
  const newContentState = Modifier.applyEntity(
    contentState,
    resizedTokenSelection,
    entityKey,
  );

  const newEditorState = EditorState.push(
    editorState,
    newContentState,
    'update-entity-selection',
  );

  const focusedSelection = selectionState.merge({
    anchorOffset: focus,
    focusOffset: focus,
  });

  // must set editorState focus to match window.getSelection() focus
  const newFocusedEditorState = EditorState.forceSelection(
    newEditorState,
    focusedSelection,
  );

  return newFocusedEditorState;
};

function tokenStrategy(contentBlock, callback, contentState) {
  contentBlock.findEntityRanges((character) => {
    const entityKey = character.getEntity();
    if (entityKey === null) return false;

    const entity = contentState.getEntity(entityKey);
    const entityType = entity.getType();

    return entityType === TOKEN_ENTITY_TYPE;
  }, callback);
}

export const highlightSelection = (newEditorState) => {
  const { start, end } = getSelectionStartAndEnd(newEditorState);
  const text = getTextFromEditorState(newEditorState);
  const token = text.substr(start, end - start);
  const newTokenSelection = { token, start, end, category: NEW_TOKEN_CATEGORY };

  const editorState = applyEntityToToken(newEditorState, newTokenSelection);
  const focusedEditorState = EditorState.forceSelection(
    editorState,
    newEditorState.getSelection(),
  );

  return focusedEditorState;
};

export const applyEntitiesToTokens = (editorState, tokens, props) =>
  tokens.reduce(
    (stateAccumulator, token) =>
      applyEntityToToken(stateAccumulator, token, props),
    editorState,
  );

const applyEntityToToken = (editorState, token, props = {}) => {
  const contentState = editorState.getCurrentContent();
  const contentStateWithEntity = contentState.createEntity(
    TOKEN_ENTITY_TYPE,
    'MUTABLE',
    { token, ...props },
  );
  const entityKey = contentStateWithEntity.getLastCreatedEntityKey();
  const { start, end } = token;
  const blockKey = contentStateWithEntity.getFirstBlock().getKey();
  const selectionState = SelectionState.createEmpty(blockKey);
  const tokenSelection = selectionState.merge({
    anchorOffset: start,
    focusOffset: end,
  });
  const newContentState = Modifier.applyEntity(
    contentStateWithEntity,
    tokenSelection,
    entityKey,
  );

  const updatedEditorState = EditorState.push(
    editorState,
    newContentState,
    TOKEN_ENTITY_TYPE,
  );
  return updatedEditorState;
};

export const shouldRenderEntities = (query, tokens = [], displayTokens) => {
  if (!displayTokens) return false;
  if (tokens.length === 0) return false;
  if (!query) return false;
  if (!queryHasToken(query, tokens)) return false;
  return true;
};

export const clearEntities = (editorState) => {
  if (!shouldClearEntities(editorState)) return editorState;

  const editorTextLength = getTextFromEditorState(editorState).length;
  const contentState = editorState.getCurrentContent();
  const blockKey = contentState.getFirstBlock().getKey();
  const selectionState = SelectionState.createEmpty(blockKey);

  const tokenSelection = selectionState.merge({
    anchorOffset: 0,
    focusOffset: editorTextLength,
  });

  const newContentState = Modifier.applyEntity(
    contentState,
    tokenSelection,
    null,
  );

  const newEditorState = EditorState.push(
    editorState,
    newContentState,
    TOKEN_ENTITY_TYPE,
  );

  const focusedEditorState = EditorState.forceSelection(
    newEditorState,
    editorState.getSelection(),
  );

  return focusedEditorState;
};

export const shouldClearEntities = (editorState) => {
  const text = getTextFromEditorState(editorState);
  if (!text) return false;

  const rawContent = convertToRaw(editorState.getCurrentContent());
  if (!Object.keys(rawContent.entityMap).length) return false;

  return true;
};

export const getTextFromEditorState = (editorState) => {
  const contentState = editorState.getCurrentContent();
  return contentState.getPlainText();
};

export const getSelectionStartAndEnd = (editorState) => {
  const selection = editorState.getSelection();
  const start = selection.getStartOffset();
  const end = selection.getEndOffset();
  return { start, end };
};

/*
  Returns location of focusOffset in draft-js search div.
   - Necessary because window.getSelection() returns focusOffset within a single Node.
   - We want the focusOffset within the parent div (draft-js search div).
*/
const getSelectionFocus = (selection, searchElement) => {
  // children of searchElement are the <span>s in the search div
  const searchChildren = searchElement.children;
  if (!searchElement.contains(selection.focusNode)) return;

  // loops through the div, adding the span's length to counter until we hit the span with focus
  let focus = 0;
  for (let index = 0; index < searchChildren.length; index++) {
    const span = searchChildren[index];
    const hasEntity = span.children.length > 1;
    const node = hasEntity ? span.children[2].children[0] : span.children[0];
    // when focus node found, we check which part of span has focus to tell where focusOffset is
    if (span.contains(selection.focusNode)) {
      // focus in left handle means add 0, focus in right handle means add entire length of span
      if (selection.focusNode.className === RIGHT_HANDLE)
        focus += node.innerHTML.length;
      else if (selection.focusNode.className !== LEFT_HANDLE)
        // focus in text means add focusOffset
        focus += selection.focusOffset;
      break;
    }
    focus += node.innerHTML.length;
  }

  return focus;
};
