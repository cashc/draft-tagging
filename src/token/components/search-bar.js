import { Component } from 'react';
import lodash from 'lodash';
const { debounce } = lodash;
import { h } from 'react-hyperscript-helpers';
import { connect } from 'react-redux';

import { HighlightedSearchTokens } from './highlighted-search-tokens';
import {
  setConnectSearchQuery,
  searchConnect,
  updateSelectedToken,
} from '../action-creators';
import { getSortedTokens } from '../selectors';
import { MIN_QUERY_LENGTH } from '../constants';

export const AUTO_SEARCH_DELAY = 2000;
export const RESUBMIT_DELAY = 2000;

const noop = () => {};

export class SearchBar extends Component {
  static defaultProps = {
    setQuery: noop,
    submitQuery: noop,
    submitQueryAfterDelay: noop,
    updateToken: noop,
    tokens: [],
    defaultQuery: '',
    disableTyping: false,
  };

  componentWillMount() {
    this.setState({ query: this.props.defaultQuery });
  }

  state = {
    query: '',
  };

  setLocalQuery = (newQuery, shouldAutoSearch = true) => {
    const { setQuery, submitQuery, submitQueryAfterDelay } = this.props;

    this.setState({ query: newQuery });
    setQuery(newQuery);

    if (newQuery.length >= MIN_QUERY_LENGTH) {
      if (shouldAutoSearch) {
        submitQueryAfterDelay(newQuery);
      } else {
        submitQuery(newQuery);
        submitQueryAfterDelay.cancel();
      }
    } else {
      submitQueryAfterDelay.cancel();
    }
  };

  render() {
    const { query } = this.state;
    const { tokens, disableTyping, updateToken } = this.props;
    return h(HighlightedSearchTokens, {
      updateToken,
      query,
      handleSearch: this.setLocalQuery,
      tokens,
      disableTyping,
    });
  }
}

function mapStateToProps(state) {
  return {
    tokens: getSortedTokens(state),
  };
}

function mapDispatchToProps(dispatch) {
  return {
    setQuery: (query) => dispatch(setConnectSearchQuery({ query })),
    submitQuery: debounce(
      (query) => {
        dispatch(searchConnect({ query }));
      },
      RESUBMIT_DELAY,
      { leading: true },
    ),
    submitQueryAfterDelay: debounce((query) => {
      dispatch(searchConnect({ query }));
    }, AUTO_SEARCH_DELAY),
    updateToken: (selectedToken, updatedToken) =>
      dispatch(
        updateSelectedToken({
          selectedToken,
          updatedToken,
        }),
      ),
  };
}

export const SearchBarConn = connect(mapStateToProps, mapDispatchToProps)(
  SearchBar,
);
