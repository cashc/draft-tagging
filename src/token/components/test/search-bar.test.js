import { h } from 'react-hyperscript-helpers';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { SearchBar } from '../../components/search-bar';

jest.useFakeTimers();

describe('when defaultQuery is supplied to SearchBar', () => {
  it('should set the query state to the defaultQuery', () => {
    const query = 'Hi there does this work?';
    const wrapper = shallow(h(SearchBar, { defaultQuery: query }), {
      experimentalLifecycle: true,
    });

    expect(wrapper.state('query')).toEqual(query);
  });
});

describe('default SearchBar render', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(h(SearchBar));
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('when disableTyping is `true`', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(h(SearchBar, { disableTyping: true }));
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('SearchBar search', () => {
  it('should not submit query when query < MIN_QUERY_LENGTH', () => {
    const setQuery = jest.fn();
    const submitQuery = jest.fn();
    const submitQueryAfterDelay = { cancel: jest.fn() };
    const wrapper = shallow(
      h(SearchBar, { setQuery, submitQuery, submitQueryAfterDelay }),
    );
    wrapper.instance().setLocalQuery('i');

    expect(setQuery).toHaveBeenCalled();
    expect(submitQueryAfterDelay.cancel).toHaveBeenCalled();
    expect(submitQuery).not.toHaveBeenCalled();
  });

  describe('autosearch set to true', () => {
    it('should submit query when query >= MIN_QUERY_LENGTH', () => {
      const setQuery = jest.fn();
      const submitQuery = jest.fn();
      const submitQueryAfterDelay = jest.fn();
      const wrapper = shallow(
        h(SearchBar, { setQuery, submitQuery, submitQueryAfterDelay }),
      );
      wrapper.instance().setLocalQuery('ivan');

      expect(setQuery).toHaveBeenCalled();
      expect(submitQueryAfterDelay).toHaveBeenCalled();
      expect(submitQuery).not.toHaveBeenCalled();
    });
  });

  describe('autosearch set to false', () => {
    it('should submit query when query >= MIN_QUERY_LENGTH', () => {
      const setQuery = jest.fn();
      const submitQuery = jest.fn();
      const submitQueryAfterDelay = { cancel: jest.fn() };
      const wrapper = shallow(
        h(SearchBar, { setQuery, submitQuery, submitQueryAfterDelay }),
      );
      wrapper.instance().setLocalQuery('ivan', false);

      expect(setQuery).toHaveBeenCalled();
      expect(submitQueryAfterDelay.cancel).toHaveBeenCalled();
      expect(submitQuery).toHaveBeenCalled();
    });
  });
});
