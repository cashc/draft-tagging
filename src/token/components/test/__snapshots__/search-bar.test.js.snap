// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default SearchBar render should match snapshot 1`] = `
<HighlightedSearchTokens
  disableTyping={false}
  handleSearch={[Function]}
  query=""
  tokens={Array []}
  updateToken={[Function]}
/>
`;

exports[`when disableTyping is \`true\` should match snapshot 1`] = `
<HighlightedSearchTokens
  disableTyping={true}
  handleSearch={[Function]}
  query=""
  tokens={Array []}
  updateToken={[Function]}
/>
`;
